### PR TITLE
Ancestor-exists() Case Query Function and Allow multiple terms case search field

### PR DESCRIPTION
**Pull request by Jtang-1**
## Product Description

This changes adds two functionalities: 
1. We can configure a case search that checks the parent case for a search field with multiple choices

https://github.com/dimagi/commcare-hq/assets/88759246/9029614f-cb3d-47b9-b0c1-4881a8111c31

2. An ancestor-exists() case search query  

https://github.com/dimagi/commcare-hq/assets/88759246/2ea70194-3ca3-4567-8c5b-246199638ed7

## Technical Summary

Jira Ticket: https://dimagi-dev.atlassian.net/browse/USH-3221
Spec: https://docs.google.com/document/d/1WbcogXJYVOV8v34NEf6ojUxHcxPtCBlFmXIKCEwFRcc/edit#heading=h.d7s2sycdfk92

## Feature Flag

Case Search

### Automated test coverage

Adds testing that ancestor-exists returns expected cases. 
Will add additional testing

### QA Plan

QA: https://dimagi-dev.atlassian.net/browse/QA-5125?search_id=d2f58eb3-a110-4b00-8b1c-65f7ec12a54f

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by Jtang-1** [comment]
Haven't yet incorporated the short circuit as described [here](https://docs.google.com/document/d/1L8dwkrH4M9GdJ51uGcd7tEG8bdAWQyenSjOEHkjKU9Y/edit#heading=h.sv0wuchw3mrg) but would like to get an initial review on the core change

**Comment by Jtang-1** [comment]
@snopoke I incorporated the short-circuit with @case_id - PR is ready to be reviewed

**Review by snopoke — COMMENTED** [review]
I haven't done a full review yet. I got stuck on the 'multi_term' piece. I was wondering if there isn't a better way to do this. What if we add a new function handler that does the 'unwrapping' for multi term values? That would allow us to provide the context on which values need unwrapping within the expression itself rather than globally in the context.

So you would change the xpath expression to `parent/prop = unwrap-list('["val1", "val2"]')`.

(note that I've also encoded the list as JSON to avoid having to do the quote replacement later).

**Comment by Jtang-1** [comment]
@snopoke I was stuck on a good way to do this too..I don't like how it's globally defined either. I like your idea and tried to implement it, but I think `parent/prop = unwrap-list('["val1", "val2"]')` has errors when parsing. From xpath spec, it seems only strings (ast.Step) are valid. 
Screen Shot 2023-05-26 at 1 32 27 PM

There might be some hacky way like doing `parent/city = "unwrap-list('['SF', 'SJ']')"`. Then [`unwrap_value`](https://github.com/dimagi/commcare-hq/blob/5c41a53625b739c7c3cb00bbafb96d7c0a964d2a/corehq/apps/case_search/dsl_utils.py#L13)  is passed the string `"unwrap-list('['SF', 'SJ']')"` as a value. Then either 1. quote replace so that it can parsed into an ast and unwrap  ['SF', 'SJ'] into a list. or 2. check if the string starts with `unwrap-list(`and ends with `)`, get the inside value. But seems very messy.

Another option might be make the xpath ` unwrap-list(parent/prop='["SF", "SJ"]') `. It's a bit of an improvement that it'smore locally defined. But I don't like how subsequent functions need to check and handle cases where the node can be either a FunctionCall or BinaryExpr
I roughly put together what that would look like here https://github.com/dimagi/commcare-hq/compare/jt/ancestor-exists...jt/ancestor-exists-unwrap-list?expand=1

**Comment by snopoke** [comment]
@Jtang-1 I don't have a problem parsing that expression

```python
from eulxml.xpath import parse as parse_xpath
parse_xpath("""parent/prop = unwrap-list('["val1", "val2"]')""")
>>> <BinaryExpression parent/prop = unwrap-list('["val1", "val2"]')>
```

There are some other alternatives:

1. Join the list with spaces and use the "selected" function instead of "ancestor-exists": `selected(parent/pro, "val1 val2 val3")`
    - This has the advantage of not having to worry about unwrapping. The downside is you can't match values with spaces in them. Generally I don't think is should be an issue but it could come up at some point.
2. Since we are only constructing this "xpath" expression on the server side for now we could make it something like `ancestor-exists(parent, 1 and selected(prop, "val1 val2 val3"))`
    - We would detect and remove the `1 and` during processing.
    - This seems roughly the same as implementing `unwrap` except that we may need to do this at some point anyway to support other use cases.

**Comment by snopoke** [comment]
> I roughly put together what that would look like here https://github.com/dimagi/commcare-hq/compare/jt/ancestor-exists...jt/ancestor-exists-unwrap-list?expand=1

This seems overly complex. You should just be able to return the unwrapped value as a list in the same way that we do for the [date](https://github.com/dimagi/commcare-hq/blob/dde833ba626268f52588637b58265e1f55472757/corehq/apps/case_search/xpath_functions/value_functions.py#L18) function.

**Comment by Jtang-1** [comment]
@snopoke Was able to get `unwrap-list` working [0675504](https://github.com/dimagi/commcare-hq/pull/33014/commits/0675504906861a24cd34cb4cde38e06816a8d9b5) [bbed552](https://github.com/dimagi/commcare-hq/pull/33014/commits/bbed5523e1381d130301f06d3f116aaabcc62f2e) I like this the best, thanks for the suggestion.

Side Question:  In the alternatives proposed,  #2 will have the same issue with spaces as #1 correct?

File: corehq/apps/case_search/xpath_functions/ancestor_functions.py [hunk 1]
```diff
@@ -23,22 +26,18 @@ def ancestor_comparison_query(context, node):
...
-    # fetch the ids of the highest level cases that match the case_property
-    # i.e. all the cases which have `property = 'value'`
-    case_ids = _parent_property_lookup(context, node)
-
     # extract ancestor path:
     # `parent/grandparent/property = 'value'` --> `parent/grandparent`
-    ancestor_path = node.left.left
-    return walk_ancestor_hierarchy(context, ancestor_path, case_ids)
+    ancestor_path = serialize(node.left.left)
+    parsed_ancestor_path = parse_xpath(ancestor_path)
```
Comment (snopoke) [inline corehq/apps/case_search/xpath_functions/ancestor_functions.py:34]: Is it necessary to serialize this and then re-parse it?

File: corehq/apps/case_search/xpath_functions/ancestor_functions.py [hunk 2]
```diff
@@ -23,22 +26,18 @@ def ancestor_comparison_query(context, node):
...
+
+    case_property = serialize(node.left.right)
+    value = serialize(node.right)
+    parsed_ancestor_filter = parse_xpath(f'{case_property}{node.op}{value}')
```
Comment (snopoke) [inline corehq/apps/case_search/xpath_functions/ancestor_functions.py:38]: nit: You could just do this:
```python
ancestor_filter = BinaryExpression(case_property, node.op, value)
```

File: corehq/apps/case_search/xpath_functions/ancestor_functions.py [hunk 3]
```diff
@@ -92,22 +91,73 @@ def _is_ancestor_path_expression(node):
...
+
+    incompatible_xpath_query_functions = {'subcase-exists', 'subcase-count'}
+
+    if isinstance(node, FunctionCall) and node.name in incompatible_xpath_query_functions:
+        raise CaseFilterError(gettext(f"{node.name} is not supported with ancestor-exists"), node.name)
+
+    if hasattr(node, 'op') and node.op in OPERATOR_MAPPING:
+        # This node is a leaf
+        return OPERATOR_MAPPING[node.op](_validate_ancestor_exists_filter(node.left),
+                                        _validate_ancestor_exists_filter(node.right))
```
Comment (snopoke) [inline corehq/apps/case_search/xpath_functions/ancestor_functions.py:137]: Other than recursively call the validation, what is the need for the operator mapping here? It doesn't appear that the return value is used at all.

File: corehq/apps/case_search/xpath_functions/ancestor_functions.py [hunk 4]
```diff
@@ -92,22 +91,73 @@ def _is_ancestor_path_expression(node):
...
+def _get_case_ids_from_ast_filter(context, filter_node):
+    from corehq.apps.case_search.dsl_utils import unwrap_value
+    if (isinstance(filter_node, BinaryExpression)
+    and serialize(filter_node.left) == "@case_id" and filter_node.op in (EQ, NEQ)):
+        # case id is provided in query i.e @case_id="b9eaf791-e427-482d-add4-2a60acf0362e"
+        case_ids = unwrap_value(filter_node.right, context)
+        if isinstance(case_ids, str):
+            return (case_id for case_id in [case_ids])
+        else:
+            return (case_id for case_id in case_ids)
```
Comment (snopoke) [inline corehq/apps/case_search/xpath_functions/ancestor_functions.py:150]: I don't think it's necessary to make these generators:
```suggestion
        return [case_ids] if isinstance(case_ids, str) else case_ids
```

File: corehq/apps/case_search/xpath_functions/value_functions.py [hunk 5]
```diff
@@ -109,3 +110,11 @@ def date_add(node, context):
         raise XPathFunctionException(str(e), serialize(node))
 
     return result.strftime(ISO_DATE_FORMAT)
+
+
+def unwrap_list(node, context):
```
Comment (snopoke) [inline corehq/apps/case_search/xpath_functions/value_functions.py:115]: 👍🏻

**Review by snopoke — COMMENTED** [review]
This is looking great. Few small comments.

**Review by snopoke — COMMENTED** [review]
This is looking great. Few small comments.

**Review by snopoke — COMMENTED** [review]
This is looking great. Few small comments.

**Review by snopoke — COMMENTED** [review]
This is looking great. Few small comments.

File: corehq/apps/case_search/filter_dsl.py [hunk 6]
```diff
@@ -26,6 +26,7 @@
 class SearchFilterContext:
     domain: str
     fuzzy: bool = False
+    multi_term: bool = False
```
[the remark this answers is not in the document]
Comment (Jtang-1) [reply corehq/apps/case_search/filter_dsl.py]: [a1c2e7b](https://github.com/dimagi/commcare-hq/pull/33014/commits/a1c2e7bde5d0c2860a7e09894d8271073ada83a7)

File: corehq/apps/case_search/xpath_functions/ancestor_functions.py (hunk 1 above)
Comment (Jtang-1) [reply corehq/apps/case_search/xpath_functions/ancestor_functions.py]: no there isn't, [6f6fa88](https://github.com/dimagi/commcare-hq/pull/33014/commits/6f6fa88a64d8fe1bd4dc5a3483a89f1ecdf086a6)

File: corehq/apps/case_search/xpath_functions/ancestor_functions.py (hunk 2 above)
Comment (Jtang-1) [reply corehq/apps/case_search/xpath_functions/ancestor_functions.py]: Ah great, I couldn't figure how to do that [6f6fa88](https://github.com/dimagi/commcare-hq/pull/33014/commits/6f6fa88a64d8fe1bd4dc5a3483a89f1ecdf086a6)

File: corehq/apps/case_search/xpath_functions/ancestor_functions.py (hunk 3 above)
Comment (Jtang-1) [reply corehq/apps/case_search/xpath_functions/ancestor_functions.py]: you're right, there is no need - [0368c6a](https://github.com/dimagi/commcare-hq/pull/33014/commits/0368c6a215ccd3c891e30c2d609997358c17899c)

File: corehq/apps/case_search/xpath_functions/ancestor_functions.py (hunk 4 above)
Comment (snopoke) [reply corehq/apps/case_search/xpath_functions/ancestor_functions.py:144]: I don't think there would need to be any type checking as long as the caller treats the returned value as an iterable.

File: corehq/apps/case_search/tests/test_ancestor_functions.py [hunk 7]
```diff
@@ -162,3 +163,15 @@ def test_nested_ancestor_exists(self):
             {'_xpath_query': xpath},
         )
         self.assertItemsEqual(query1.get_ids(), ['a1'])
+
+    @generate_cases([
+        ("ancestor_exists(status='active' and subcase-exists('parent', city = 'LA'))",),
+        ("ancestor_exists(status='active' and subcase-count('parent', city = 'LA')) > 3",),
```
Comment (snopoke) [inline corehq/apps/case_search/tests/test_ancestor_functions.py:169]: I think these should be `ancestor-exists` (dash not underscore)

**Review by snopoke — APPROVED** [review]
Pending latest comment about test cases

**Review by snopoke — APPROVED** [review]

File: corehq/apps/case_search/tests/test_ancestor_functions.py (hunk 7 above)
Comment (Jtang-1) [reply corehq/apps/case_search/tests/test_ancestor_functions.py]: thanks! [9eb4fc1](https://github.com/dimagi/commcare-hq/pull/33014/commits/9eb4fc14e79fbd181fcff0bcdd0d4abc4acef36e)

**Pull request merged**